### PR TITLE
fix(#8777): IndexedDB polyfill

### PR DESCRIPTION
**Pull request by latin-panda**
# Description

- Adds service to return a list of db names when `indexedDB.databases()` isn't supported by the browser --- looks at Firefox 😒

#8777

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs

If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:indexdb-polyfill-firefox/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:indexdb-polyfill-firefox/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:indexdb-polyfill-firefox/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by latin-panda** [comment]
@lorerod Here is the process; it still needs more manual testing; I'll do it on Monday, but I am opening this in case you have the chance to take a look and provide feedback. I tried to minimize the changes in the Telemetry service since we expect to remove this polyfill once Firefox adds support to this.

**Comment by latin-panda** [comment]
@lorerod @Benmuiruri, this is ready for review. Can you please have a look? Telemetry should work in Chrome and Firefox as usual.

File: webapp/tests/karma/ts/services/indexed-db.service.spec.ts
Comment (latin-panda) [inline webapp/tests/karma/ts/services/indexed-db.service.spec.ts:1]: I've created an issue to remove this in the future
https://github.com/medic/cht-core/issues/8787

File: webapp/src/ts/services/indexed-db.service.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+import { DbService } from '@mm-services/db.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class IndexedDbService {
+  private readonly LOCAL_DOC = '_local/indexeddb-placeholder';
+  private loadingLocalDoc: Promise<IndexedDBDoc> | undefined;
+  private hasDatabasesFn: boolean;
```
Comment (Benmuiruri) [inline webapp/src/ts/services/indexed-db.service.ts:12]: Nice  usage👌

File: webapp/src/ts/services/indexed-db.service.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    private dbService: DbService,
+    @Inject(DOCUMENT) private document: Document
+  ) {
+    // Firefox doesn't support the databases() function. Issue: https://github.com/medic/cht-core/issues/8777
+    this.hasDatabasesFn = !!this.document.defaultView?.indexedDB?.databases;
+  }
+
+  async getDatabaseNames() {
+    if (this.hasDatabasesFn) {
+      const databases = await this.document.defaultView!.indexedDB.databases();
```
Comment (Benmuiruri) [inline webapp/src/ts/services/indexed-db.service.ts:25]: The check `if (this.hasDatabasesFn)` makes it reasonable to assume that indexedDB.databases is available hence it is safe to use the ! operator.  But from a defensive programming view, perhaps using `const databases = await this.document.defaultView?.indexedDB?.databases();` would be more robust.

**Review by Benmuiruri — APPROVED** [review]
LGTM! Just one thought I have shared.

File: webapp/tests/karma/ts/services/indexed-db.service.spec.ts
Comment (lorerod) [inline webapp/tests/karma/ts/services/indexed-db.service.spec.ts:1]: Thanks for adding these test scenarios!

**Review by lorerod — APPROVED** [review]

File: webapp/src/ts/services/indexed-db.service.ts (hunk 2 above)
Comment (latin-panda) [reply webapp/src/ts/services/indexed-db.service.ts]: yes, I was back and forth on this. I applied the suggestion

**Pull request merged**